synapse/handlers/user_directory.py

```diff
@@ -490,5 +479,11 @@ async def _handle_possible_remote_profile_change(
         if not isinstance(new_avatar, str):
             new_avatar = prev_avatar
 
-        if prev_name != new_name or prev_avatar != new_avatar:
+        if (
+            prev_name != new_name
+            or prev_avatar != new_avatar
+            or prev_event_id == event_id
+        ):
+            # Only update if something has changed, or we didn't have a previous event
+            # in the first place.
             await self.store.update_profile_in_user_dir(user_id, new_name, new_avatar)
```

Comment:
When there is no prev event, we can now try to update the user directory with non-strings, despite the type checks above.

Reply:
Hm this was already flawed by the looks: you could just send a dodgy event and then follow it up with another dodgy event to get the `prev_event` accepted.

Reply:
But `update_profile_in_user_dir` sets it to None if you pass in a non-string, so I suggest we just replace non-strings with None anyway.

Reply:
That approach sounds reasonable!